Implement updated TextTrackCue constructor proposal

marcoscaceres opened the pull request:
#### ed93c318ea694da22ab23080a68b7f6240ed997f
<pre>
Implement updated TextTrackCue constructor proposal
<a href="https://bugs.webkit.org/show_bug.cgi?id=262053">https://bugs.webkit.org/show_bug.cgi?id=262053</a>
rdar://116002871

Reviewed by Eric Carlson.

Implements spec changes suggested at:
<a href="https://github.com/whatwg/html/pull/9771">https://github.com/whatwg/html/pull/9771</a>

Specifically:

 * supports `b` and `i` elements, to match WebVTT
 * drops support for allowing `img` in the cue fragment
 * implements more robust error checking to prevent malformed ::cue and :cuebackground hierarchies
 * As such, ::tagPseudoObjects() can now throw
 * Switch from InvalidStateError to InvalidNodeTypeError for when required attributes are missing
 * implements cuebackground and cue global attributes

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/reactions/HTMLElement-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/reactions/HTMLElement.html:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/cue-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/cue.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/cuebackground-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/cuebackground.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/historical-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/historical.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/interfaces/TextTrackCue/constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/interfaces/TextTrackCue/constructor.html:
* LayoutTests/inspector/model/remote-object/dom-expected.txt:
* LayoutTests/media/track/texttrackcue/texttrackcue-constructor-expected.txt:
* LayoutTests/media/track/texttrackcue/texttrackcue-constructor.html:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLElement.idl:
* Source/WebCore/html/track/TextTrackCue.cpp:
(WebCore::isLegalNode):
(WebCore::tagPseudoObjects):
(WebCore::removePseudoAttributes):
(WebCore::TextTrackCue::create):
(WebCore::TextTrackCue::setPauseOnExit):
(WebCore::cueAttributName): Deleted.
(WebCore::cueBackgroundAttributName): Deleted.

Canonical link: <a href="https://commits.webkit.org/268644@main">https://commits.webkit.org/268644@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3285664e4966ce19f4dbc582bed3da79c8a9a299

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20139 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20576 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21197 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22040 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18798 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20370 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23816 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20737 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20261 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20359 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20266 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17500 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22889 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17436 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18297 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24547 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18513 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18473 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22540 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19071 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16187 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18271 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18133 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4867 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22613 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18896 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/html/track/TextTrackCue.cpp:
```diff
@@ -65,13 +65,13 @@ namespace WebCore {
+static const QualifiedName& cueBackgroundAttributeName()
 {
     static NeverDestroyed<QualifiedName> cueBackgroundTag(nullAtom(), "cuebackground"_s, nullAtom());
     return cueBackgroundTag;
```
Should we instead add these to `Source/WebCore/html/HTMLAttributeNames.in`? Presumably we need IDL attributes for them as well and such?

annevk replied:
I think you should be able to get rid of these now that you added them there.

marcoscaceres replied:
Ok, will try removing them.

annevk commented on Source/WebCore/html/HTMLElement.idl:
```diff
@@ -34,6 +34,10 @@
 
+    // TextTrackCue related
+    [Reflect] attribute boolean cue;
+    [Reflect] attribute boolean cuebackground;
```
You need CEReactions=Needed for both.

marcoscaceres replied:
I was wondering about that. I guess it would make sense just to have them generically even if they are not really intended to be used that way.

annevk replied:
Yeah, otherwise modifications of these attributes wouldn't trigger callbacks which would violate custom elements. (I will grant that it's unlikely that anyone needs this, but that's not a good enough reason to introduce inconsistencies.)

marcoscaceres replied:
Agree. Added.

eric-carlson commented on LayoutTests/media/track/texttrackcue/texttrackcue-constructor-expected.txt:
```diff
@@ -2,7 +2,7 @@
-PASS TextTrackCue should throw "InvalidStateError" if documentFragment doesn't have 'cuebackground' and 'cue' nodes
-PASS TextTrackCue should allow all valid node types
+PASS TextTrackCue should throw "InvalidNodeTypeError" if documentFragment doesn't have 'cuebackground' and 'cue' nodes
+FAIL TextTrackCue should allow all valid node types Invalid node type: IMG
```
We should remove `img` from the list of valid node types in this test

marcoscaceres replied:
oh oops! well spotted!

annevk commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -776,6 +776,28 @@ void HTMLElement::setTranslate(bool enable)
+bool HTMLElement::cue() const
+{
+    const AtomString& value = attributeWithoutSynchronization(cueAttr);
+    return equalLettersIgnoringASCIICase(value, "true"_s) ? true : false;
```
Sorry for not noticing this sooner, but why are we not making these boolean attributes? True if present, false if absent.

annevk replied:
As in, just return `hasAttributeWithoutSynchronization()` here and either add with the empty string or remove below.

marcoscaceres replied:
Oh yeah, right! oops. I'll a test for this also.

annevk commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -776,6 +776,33 @@ void HTMLElement::setTranslate(bool enable)
+        setAttributeWithoutSynchronization(cueAttr, emptyAtom());
+    else
+        removeAttribute(cueAttr);
+}
```
This is still wrong. However, are you sure we actually need this? I think `[Reflect]` might do it as there's nothing for the `hidden` attribute either for instance.

annevk commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -776,6 +776,33 @@ void HTMLElement::setTranslate(bool enable)
+
+bool HTMLElement::cuebackground() const
+{
+
```
Nit: remove this newline.

marcoscaceres committed:
Implement updated TextTrackCue constructor proposal
https://bugs.webkit.org/show_bug.cgi?id=262053
rdar://116002871

Reviewed by Eric Carlson.

Implements spec changes suggested at:
https://github.com/whatwg/html/pull/9771

Specifically:

 * supports `b` and `i` elements, to match WebVTT
 * drops support for allowing `img` in the cue fragment
 * implements more robust error checking to prevent malformed ::cue and :cuebackground hierarchies
 * As such, ::tagPseudoObjects() can now throw
 * Switch from InvalidStateError to InvalidNodeTypeError for when required attributes are missing
 * implements cuebackground and cue global attributes

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/reactions/HTMLElement-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/reactions/HTMLElement.html:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/cue-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/cue.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/cuebackground-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/cuebackground.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/historical-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/historical.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/interfaces/TextTrackCue/constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/interfaces/TextTrackCue/constructor.html:
* LayoutTests/inspector/model/remote-object/dom-expected.txt:
* LayoutTests/media/track/texttrackcue/texttrackcue-constructor-expected.txt:
* LayoutTests/media/track/texttrackcue/texttrackcue-constructor.html:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLElement.idl:
* Source/WebCore/html/track/TextTrackCue.cpp:
(WebCore::isLegalNode):
(WebCore::tagPseudoObjects):
(WebCore::removePseudoAttributes):
(WebCore::TextTrackCue::create):
(WebCore::TextTrackCue::setPauseOnExit):
(WebCore::cueAttributName): Deleted.
(WebCore::cueBackgroundAttributName): Deleted.

Canonical link: https://commits.webkit.org/268644@main
